src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -154,29 +164,63 @@ void mir::frontend::XdgDecorationManagerV1::get_toplevel_decoration(wl_resource*
 }
 
 mir::frontend::XdgToplevelDecorationV1::XdgToplevelDecorationV1(
-    wl_resource* id, mir::frontend::XdgToplevelStable* toplevel) :
+    wl_resource* id, mir::frontend::XdgToplevelStable* toplevel, std::shared_ptr<DecorationStrategy> strategy) :
     wayland::XdgToplevelDecorationV1{id, Version<1>{}},
-    toplevel{toplevel}
+    toplevel{toplevel},
+    decoration_strategy{strategy}
 {
 }
 
-void mir::frontend::XdgToplevelDecorationV1::update_mode(uint32_t new_mode)
+auto mir::frontend::XdgToplevelDecorationV1::decorations_type_to_protocol_mode(DecorationStrategy::DecorationsType type)
+    -> uint32_t
 {
-    auto spec = shell::SurfaceSpecification{};
+    switch (type)
+    {
+    case DecorationStrategy::DecorationsType::ssd:
+        return Mode::server_side;
+    case DecorationStrategy::DecorationsType::csd:
+        return Mode::client_side;
+    default:
+        mir::log_warning("%s: Got invalid protocol decorations type, defaulting to client side", __FUNCTION__);
+        return Mode::client_side;
+    }
+}
 
-    mode = new_mode;
+auto mir::frontend::XdgToplevelDecorationV1::protocol_mode_to_decorations_type(uint32_t mode)
+    -> DecorationStrategy::DecorationsType
+{
     switch (mode)
     {
     case Mode::client_side:
-        spec.server_side_decorated = false;
-        break;
+        return DecorationStrategy::DecorationsType::csd;
     case Mode::server_side:
+        return DecorationStrategy::DecorationsType::ssd;
+    default:
+        mir::log_warning("%s: Got invalid protocol mode (%d), defaulting to client side", __FUNCTION__, mode);
```

Comment:
A warning log message should explain what the client did wrong (e.g. what the `new_mode` was and, ideally, which client). I don't think logging the function name adds to that explanation.